pallets/gear/src/internal.rs

```diff
@@ -771,8 +774,23 @@ where
             Some(hold.expected())
         } else {
             // Permanently transferring funds.
+            // Note that we have no guarantees of the user account to exist. Since no minimum
+            // transfer value is enforced, the transfer can fail. Handle it gracefully.
             CurrencyOf::<T>::transfer(&from, &to, value, ExistenceRequirement::AllowDeath)
-                .unwrap_or_else(|e| unreachable!("Failed to transfer value: {:?}", e));
+                .unwrap_or_else(|e| match e {
+                    DispatchError::Token(TokenError::BelowMinimum) => {
+                        log::warn!(
+                            "Failed to transfer value: {:?}. User account balance is too low.",
+                            e
+                        );
+                        <CurrencyOf<T> as fungible::Unbalanced<_>>::handle_dust(fungible::Dust(
+                            value,
+                        ));
+                    }
+                    _ => {
```

Comment:
add all of the variants to pattern matching. I think we can handle dust for all case, and for ones that are not BelowMinimum throw log::error instead of panics here?

Reply:
The problem here is that the error type is `sp_runtime::DispatchError` which has tons of variants, of which the `CurrencyOf::<T>::transfer()` can only return a few. Looks impractical (if at all possible) to list all possible variants.

Regarding the second part (not panicking but logging an error), it depends on the invariants we want to uphold. If there is a value a user hopes to receive back from a program (which can easily be big), but a program can't transfer it despite all the safeguards we have in the code, it's a reason to look into it rather than just letting go, isn't it?